x-pack/plugins/cloud/public/custom_events_system.ts

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { customEvents } from '@kbn/custom-events';
+import { CoreSetup, IBasePath, PackageInfo, PluginInitializerContext } from 'kibana/public';
+import { Subscription } from 'rxjs';
+import {
+  AuthenticatedUser,
+  AuthenticationServiceStart,
+} from '../../../../x-pack/plugins/security/public';
+import { CloudConfigType } from '.';
+
+/**
+ * This is the entry point used to boot the frontend when serving a application
+ * that lives in the Kibana Platform.
+ */
+
+interface SetupDeps {
+  authc?: AuthenticationServiceStart;
+  esOrgId: string;
+}
+
+export class CustomEventsSystem {
+  private readonly enabled?: boolean;
+  private basePath!: IBasePath;
+  private appSubscription?: Subscription;
+  private packageInfo: Readonly<PackageInfo>;
+  /**
+   * `apmConfig` would be populated with relevant APM RUM agent
+   * configuration if server is started with elastic.apm.* config.
+   */
+  constructor(
+    initializerContext: PluginInitializerContext,
+    private readonly config: CloudConfigType
+  ) {
+    this.packageInfo = initializerContext.env.packageInfo;
+    this.enabled = config.full_story?.enabled;
+  }
+
+  setup(coreSetup: CoreSetup, deps: SetupDeps) {
+    if (!this.enabled) return;
+
+    this.basePath = coreSetup.http.basePath;
+
+    this.setupCustomEventReporting(coreSetup, deps).catch((e) =>
+      // eslint-disable-next-line no-console
+      console.debug(`Error setting up FullStory: ${e.toString()}`)
+    );
+  }
+
+  private async setupCustomEventReporting(coreSetup: CoreSetup, { authc, esOrgId }: SetupDeps) {
+    const { enabled, org_id: fullstoryOrgId } = this.config.full_story || {};
+
+    if (!enabled || !fullstoryOrgId) {
+      return; // do not load any fullstory code in the browser if not enabled
+    }
+    // Very defensive try/catch to avoid any UnhandledPromiseRejections
+    try {
+      const applicationPromise = coreSetup.getStartServices().then(([coreStart]) => {
+        return coreStart.application;
+      });
+
+      const userIdPromise: Promise<string | undefined> = authc
+        ? loadUserId({ getCurrentUser: authc.getCurrentUser })
+        : Promise.resolve(undefined);
+
+      const initSuccess = await customEvents.initialize({
+        fullstoryOrgId,
+        esOrgId,
+        enabled,
+        userIdPromise,
+        basePath: this.basePath,
+        packageInfo: this.packageInfo,
+      });
```

Comment:
I totally agree with @lukeelmers comments. It seems we're just getting coupled tighter to fullstory with the changes of this PR, when the goal is supposed to be the opposite. 

With a correct abstraction / delegation model, I would expect something like:


```ts
import { createCustomEvents, CustomEventDelegate } from '@kbn/custom-events';
import { initializeFullStory } from 'somewhere-else';

const fullStory: CustomEventDelegate = await initializeFullStory({ fullstoryOrgId, ...otherInitParams }); 
const customEvents = createCustomEvents({ delegate: fullStory });
```

Reply:
++ Pierre's delegation model here is similar to what I had in mind too. Except I could see potentially having multiple delegates `createCustomEvents({ delegates: [fullStory, ...futureStuff] });`